spec/query_spec.rb

```diff
@@ -62,29 +40,122 @@
+          'proxied query' => 'SELECT * FROM "t" WHERE ("x" < 1)',
+          'readyset supported' => 'pending',
+          'count' => 5,
+        ),
+      ]
+    end
+
+    before do
+      allow(Readyset::Query).to receive(:all_seen_but_not_cached).and_return(queries)
+    end
+
+    context 'when every Readyset::Query#cache! invocation succeeds' do
+      before do
+        queries.each { |query| allow(query).to receive(:cache!).with(always: true) }
+        subject
+      end
+
+      it 'invokes Readyset::Query#cache! on every supported query with the given "always" ' \
+        'parameter' do
+        supported_queries.each do |query|
+          expect(query).to have_received(:cache!).with(always: true)
+        end
+      end
+
+      it 'does not invoke Readyset::Query#cache! on any unsupported or pending queries' do
+        unsupported_or_pending_queries.each do |query|
+          expect(query).not_to have_received(:cache!)
+        end
+      end
+
+      it 'returns nil' do
+        is_expected.to be_nil
+      end
+    end
+
+    context 'when one of the Readyset::Query#cache! invocations fails' do
+      before do
+        allow(queries[0]).to receive(:cache!).and_raise(StandardError)
+        allow(queries[1]).to receive(:cache!)
+
+        begin
+          subject
+        rescue StandardError
+          nil
+        end
+      end
+
+      it 'raises the error raised by the Readyset::Query#cache! invocation' do
+        expect { subject }.to raise_error(StandardError)
+      end
+
+      it 'invokes Readyset::Query#cache! on the queries in the list up to and including the ' \
+        'query that caused the error with the given "always" parameter' do
+        expect(queries[0]).to have_received(:cache!).with(always: true)
+      end
+
+      it 'does not invoke Readyset::Query#cache! on any of the queries in the list after the ' \
+        'query that caused the error' do
+        expect(queries[1]).not_to have_received(:cache!)
+      end
+    end
+  end
+
+  describe '.drop_all_caches!' do
+    subject { Readyset::Query.drop_all_caches! }
+
+    before do
+      allow(Readyset).to receive(:raw_query).with('DROP ALL CACHES')
+
+      subject
+    end
+
+    it 'invokes "DROP ALL CACHES" against Readyset' do
+      expect(Readyset).to have_received(:raw_query).with('DROP ALL CACHES')
+    end
+
+    it 'returns nil' do
+      is_expected.to be_nil
     end
   end
 
   describe '.find' do
     subject { Readyset::Query.find(query_id) }
 
-    let(:query_id) { 'q_eafb620c78f5b9ac' }
-
     context 'when a cached query with the given ID exists' do
-      let(:query) do
-        {
-          'query id' => query_id,
-          'query text' => 'SELECT * FROM "t" WHERE ("x" = $1)',
-          'cache name' => query_id,
-          'fallback behavior' => 'fallback allowed',
-          'count' => 5,
-        }
-      end
+      let(:query) { cached_query_attributes }
 
       before do
         allow(Readyset).
```

Comment:
I found [this usage of `shared_examples`](https://github.com/thoughtbot/shoulda-matchers/blob/3654e5ca7fb10dc84a921e7f06b7875eaf8729c3/spec/unit/shoulda/matchers/active_record/have_db_index_matcher_spec.rb#L5) in the shoulda-matcher gem. 
 
 Given the repeated pattern of `setup` and `assert` instances, would something like the following help?
```ruby
  shared_examples 'query invocation' do |method, command|
    subject { send(method) }

    before do
      allow(Readyset).to receive(:raw_query).with(command).and_return(return_value)
      subject
    end

    it "invokes \"#{command}\" on Readyset" do
      expect(Readyset).to have_received(:raw_query).with(command)
    end
  end
  ```
 
 Admittedly, it was difficult for me to read the spec and understand what was going on. Your previous refactor did help a lot though! I think the coverage is excellent with *what* we're testing, but the readability in a spec should also indicate *why*.

Here's how I interpreted the `.all_seen_but_not_cached` spec after careful reading.
 ```ruby
   describe '.all_seen_but_not_cached' do
    # what we're actually executing
    subject { Readyset::Query.all_seen_but_not_cached }
    # prepare what we think the ReadySet *server* response should be. (JSON?)
    let(:seen_but_not_cached_queries) { [seen_but_not_cached_query_attributes] }

    before do
      # we are stubbing Readyset.raw_query and its response 
    end

    it 'invokes "SHOW PROXIED QUERIES" on Readyset' do
      # expect the stub to receive the right argument 
    end
```

I feel that that spec could be summarized with `it invokes SQL_COMMAND against ReadySet`. I found this [neat StackOverflow post](https://stackoverflow.com/questions/19556296/whats-the-difference-between-include-examples-and-it-behaves-like) regarding usage of `describe`, `it`, and `context`, on top of `include_examples`. 

---
**Note**:
From what I've read so far, testing should aim for *readability*, *maintainability*, and *correctness*. I am unsure if what was said above achieves that.

Here's a [neat thoughtbot article](https://thoughtbot.com/blog/the-self-contained-test) that kind of goes against `shared_examples`. 

*A quote from the article:*
> Next time you find yourself writing the same setup over and over, think about it like you’re writing a story, a piece of prose for a future developer. That future developer might even be you, and they will thank you for it.

Reply:
This is _amazing_ feedback -- thank you for the great suggestions, @helpotters! I will take a look at `shared_examples` and think carefully about how I can improve upon the readability of these specs 😄